### PR TITLE
feat: add streamlits to sdk

### PR DESCRIPTION
**Pull request by sfc-gh-swinkler**
Adds Streamlit to SDK
https://docs.snowflake.com/en/sql-reference/sql/create-streamlit

### CODE REVIEW

File: pkg/sdk/streamlits_validations_gen.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		errs = append(errs, ErrInvalidObjectIdentifier)
+	}
+	return JoinErrors(errs...)
+}
+
+func (opts *ShowStreamlitOptions) validate() error {
+	if opts == nil {
+		return ErrNilOptions
+	}
+	var errs []error
```
Comment (sfc-gh-jcieslak) [inline pkg/sdk/streamlits_validations_gen.go:54]: aren't there any validations from predefined objects? (Like, In, From)

File: pkg/sdk/testint/streamlits_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+func TestInt_Streamlits(t *testing.T) {
+	client := testClient(t)
+	ctx := testContext(t)
+
+	databaseTest, schemaTest := testDb(t), testSchema(t)
+
+	cleanupStreamlitHandle := func(id sdk.SchemaObjectIdentifier) func() {
+		return func() {
+			err := client.Streamlits.Drop(ctx, sdk.NewDropStreamlitRequest(id))
```
Comment (sfc-gh-jcieslak) [inline pkg/sdk/testint/streamlits_integration_test.go:21]: you can use IfExists instead of checking the error

File: pkg/sdk/streamlits_gen.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+type streamlitsRow struct {
+	CreatedOn      string         `db:"created_on"`
+	Name           string         `db:"name"`
+	DatabaseName   string         `db:"database_name"`
+	SchemaName     string         `db:"schema_name"`
+	Title          sql.NullString `db:"title"`
+	Owner          string         `db:"owner"`
+	Comment        sql.NullString `db:"comment"`
+	QueryWarehouse sql.NullString `db:"query_warehouse"`
+	UrlId          string         `db:"url_id"`
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/streamlits_gen.go:74]: If we have terse, I think all fields besides created_on, name, database_name, schema_name, and url_id should be nullable. Please add an integration test with terse to validate it.

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: pkg/sdk/testint/streamlits_integration_test.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+		_, err = client.Streamlits.ShowByID(ctx, id)
+		require.ErrorIs(t, err, collections.ErrObjectNotFound)
+
+		o, err := client.Streamlits.ShowByID(ctx, nid)
+		require.NoError(t, err)
+		require.Equal(t, nid.Name(), o.Name)
+	})
+
+	t.Run("show streamliit: with like", func(t *testing.T) {
```
Comment (sfc-gh-jcieslak) [inline pkg/sdk/testint/streamlits_integration_test.go:113]: "streamlit"

**Review by sfc-gh-jcieslak — CHANGES_REQUESTED** [review]

File: pkg/sdk/testint/streamlits_integration_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sfc-gh-swinkler) [reply pkg/sdk/testint/streamlits_integration_test.go]: fixed

File: pkg/sdk/streamlits_gen.go (hunk 3 above)
Comment (sfc-gh-swinkler) [reply pkg/sdk/streamlits_gen.go:74]: added terse int test

File: pkg/sdk/streamlits_def.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		Create().
+		OrReplace().
+		SQL("STREAMLIT").
+		IfNotExists().
+		Name().
+		TextAssignment("ROOT_LOCATION", g.ParameterOptions().SingleQuotes().Required()).
+		TextAssignment("MAIN_FILE", g.ParameterOptions().SingleQuotes().Required()).
+		OptionalIdentifier("Warehouse", g.KindOfT[AccountObjectIdentifier](), g.IdentifierOptions().Equals().SQL("QUERY_WAREHOUSE")).
+		OptionalTextAssignment("COMMENT", g.ParameterOptions().SingleQuotes()).
+		WithValidation(g.ValidIdentifier, "name").
```
[the remark this answers is not in the document]
Comment (sfc-gh-swinkler) [reply pkg/sdk/streamlits_def.go:30]: added validation

File: pkg/sdk/streamlits_def.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
+package sdk
+
+import g "github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk/poc/generator"
+
+//go:generate go run ./poc/main.go
+
+var streamlitSet = g.NewQueryStruct("StreamlitSet").
+	OptionalTextAssignment("ROOT_LOCATION", g.ParameterOptions().SingleQuotes().Required()).
+	OptionalTextAssignment("MAIN_FILE", g.ParameterOptions().SingleQuotes().Required()).
+	OptionalIdentifier("Warehouse", g.KindOfT[AccountObjectIdentifier](), g.IdentifierOptions().Equals().SQL("QUERY_WAREHOUSE")).
```
[the remark this answers is not in the document]
Comment (sfc-gh-swinkler) [reply pkg/sdk/streamlits_def.go:10]: added validation

File: pkg/sdk/streamlits_def.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+		Alter().
+		SQL("STREAMLIT").
+		IfExists().
+		Name().
+		OptionalQueryStructField(
+			"Set",
+			streamlitSet,
+			g.KeywordOptions().SQL("SET"),
+		).
+		Identifier("RenameTo", g.KindOfTPointer[SchemaObjectIdentifier](), g.IdentifierOptions().SQL("RENAME TO")).
```
[the remark this answers is not in the document]
Comment (sfc-gh-swinkler) [reply pkg/sdk/streamlits_def.go:45]: added validation

File: pkg/sdk/streamlits_def.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	"https://docs.snowflake.com/en/sql-reference/sql/drop-streamlit",
+	g.NewQueryStruct("DropStreamlit").
+		Drop().
+		SQL("STREAMLIT").
+		IfExists().
+		Name().
+		WithValidation(g.ValidIdentifier, "name"),
+).ShowOperation(
+	"https://docs.snowflake.com/en/sql-reference/sql/show-streamlits",
+	g.DbStruct("streamlitsRow").
```
[the remark this answers is not in the document]
Comment (scottwinkler) [reply pkg/sdk/streamlits_def.go:59]: docs not right. Posted in the channel

[the remark this answers is not in the document]
Comment (sfc-gh-swinkler) [reply pkg/sdk/streamlits_def.go:59]: docs not right, posted in channel

File: pkg/sdk/streamlits_gen_test.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+
+	t.Run("validation: incorrect identifier", func(t *testing.T) {
+		opts := defaultOpts()
+		opts.name = NewSchemaObjectIdentifier("", "", "")
+		assertOptsInvalidJoinedErrors(t, opts, ErrInvalidObjectIdentifier)
+	})
+
+	t.Run("all options", func(t *testing.T) {
+		opts := defaultOpts()
+		assertOptsValidAndSQLEquals(t, opts, `DROP STREAMLIT %s`, id.FullyQualifiedName())
```
[the remark this answers is not in the document]
Comment (sfc-gh-swinkler) [reply pkg/sdk/streamlits_gen_test.go]: added

File: pkg/sdk/streamlits_gen_test.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+	})
+
+	t.Run("show with like", func(t *testing.T) {
+		opts := defaultOpts()
+		opts.Like = &Like{
+			Pattern: String("pattern"),
+		}
+		assertOptsValidAndSQLEquals(t, opts, `SHOW STREAMLITS LIKE 'pattern'`)
+	})
+
```
[the remark this answers is not in the document]
Comment (sfc-gh-swinkler) [reply pkg/sdk/streamlits_gen_test.go:135]: added

File: pkg/sdk/streamlits_validations_gen.go (hunk 1 above)
Comment (sfc-gh-swinkler) [reply pkg/sdk/streamlits_validations_gen.go:65]: added validation for like

File: pkg/sdk/testint/streamlits_integration_test.go (hunk 2 above)
Comment (sfc-gh-swinkler) [reply pkg/sdk/testint/streamlits_integration_test.go]: good idea

File: pkg/sdk/testint/streamlits_integration_test.go (hunk 4 above)
Comment (sfc-gh-swinkler) [reply pkg/sdk/testint/streamlits_integration_test.go]: fixed

File: pkg/sdk/testint/streamlits_integration_test.go [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+		require.Equal(t, 1, len(streamlits))
+		require.Equal(t, *e, streamlits[0])
+	})
+
+	t.Run("show streamlit: terse with like", func(t *testing.T) {
+		stage, cleanupStage := createStage(t, client, databaseTest, schemaTest, random.AlphaN(4))
+		t.Cleanup(cleanupStage)
+		e := createStreamlitHandle(t, stage, "manifest.yml")
+
+		streamlits, err := client.Streamlits.Show(ctx, sdk.NewShowStreamlitRequest().WithLike(&sdk.Like{Pattern: &e.Name}))
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/testint/streamlits_integration_test.go:125]: Terse param is not set here, please set it and re-run the test to check.

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: pkg/sdk/testint/streamlits_integration_test.go (hunk 11 above)
Comment (sfc-gh-swinkler) [reply pkg/sdk/testint/streamlits_integration_test.go]: set terse, make sure that all values being set are set, and empy are empty
```
--- PASS: TestInt_Streamlits (1.06s)
    --- PASS: TestInt_Streamlits/show_streamlit:_terse_with_like (1.06s)
PASS
```

**Review by sfc-gh-asawicki — APPROVED** [review]

**Review by sfc-gh-jcieslak — APPROVED** [review]

**Pull request merged**